File: leetcode/binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.rb
# LeetCode #107 - Binary Tree Level Order Traversal II
# https://leetcode.com/problems/binary-tree-level-order-traversal-ii/

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[][]}
def level_order_bottom(root)
  out = []
  queue = []
  queue << root if root

  until queue.empty?
    row = []

    queue.size.times do
      node = queue.shift
      row << node.val
      queue << node.left if node.left
      queue << node.right if node.right
    end

    out << row
  end

  out.reverse!
end

File: leetcode/binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.java
// LeetCode #107 - Binary Tree Level Order Traversal II
// https://leetcode.com/problems/binary-tree-level-order-traversal-ii/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> rows = new ArrayList<>();
        if (root == null) return rows;

        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            List<Integer> row = new ArrayList<>();
            rows.add(row);

            for (var breadth = queue.size(); breadth > 0; --breadth) {
                var node = queue.remove();
                row.add(node.val);
                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
            }
        }

        Collections.reverse(rows);
        return rows;
    }
}
